### ISSUE TITLE
Logging may hang if called async then sync

### ISSUE THREAD

**Issue by alakoring** [CocoaLumberjack/CocoaLumberjack#507]
I put some notes in case #61 but figure this is an ongoing issue…

Running in iOS 8.3 Simulator if that matters. Downloaded Lumberjack a few days ago (I don't see its version anywhere obvious). I added a custom log style (the documentation on this is out of date but I don't think I did anything that would break.)

Things were fine until I decided to add my first `DDLogError` to my app. I now get random but repeatable deadlock. Apparently it happens when there is a pending async log generated by one of my threads, and I try a `DDLogError` (which is synchronous). The stack traces should show this.

![screen shot 2015-05-03 at 11 22 08](https://cloud.githubusercontent.com/assets/12171257/7445544/d49a4394-f186-11e4-8bad-0f225008d8aa.png)

(actually it shows the deadlock on a `flushLog` because I was experimenting with working around the deadlock. Same difference.)

**Comment by rivera-ernesto** [comment]
You can see the version in the CocoaLumberjack.podspec file.

**Comment by alakoring** [comment]
`s.version  = '2.0.0'` (believe it or not, some of us don’t use Cocoapods)

**Comment by rivera-ernesto** [comment]
I know, but the podspec file is there.

**Comment by alakoring** [comment]
Not trying to be snarky, I had no idea an implementation detail of a feature I don’t use would be important.

Also, I think I mis-categorized this bug. The stack trace shows that the DDLog call was made from the main thread. So this is apparently a problem when the async log is in progress, and a synchronous one happens. I believe both from the same time through the run loop.

**Comment by carllindberg** [comment]
It looks like the core DDLog thread is waiting on one of the actual individual logger threads to finish (presumably one of the threads where the backtrace is not expanded), and the main thread is waiting on the DDLog thread.  I'm not sure those two are deadlocks in and of themselves; which loggers are configured and what do they do with the messages?  The logger thread is calling dispatch_group_wait(), which indicates that this is a multiprocessor machine and the individual loggers are invoked on separate threads of their own, but one of them is not completing, perhaps because it is waiting on one of the first two threads.  But we'd need to see that backtrace too.

**Comment by Elanz** [comment]
Hello,

We also hit what we think is exactly this deadlock, our stack trace looks very similar to the one David posted:

![screen shot 2015-06-24 at 3 10 23 pm](https://cloud.githubusercontent.com/assets/856806/8343413/e221c510-1a89-11e5-8eb1-f6b6c3c7ea4c.png)

We believe that the issue in our specific case is that we are calling DDLogError (a macro that is always synchronous) from within the thread which is rolling the log file.  Doing experiments to try and prove that this is our bug, but wanted to let you know that while this is extremely rare it does seem like a valid deadlock.

**Comment by carllindberg** [comment]
Yes, invoking the logging macros within Logger implementations is generally a bad idea.  Any synchronous log will need to synchronously invoke the logger itself again, so that is an automatic deadlock.  I think most of the existing logger implementations have simple macros which use NSLog for any internal logging needs.  If you have to log, you should probably ensure async no matter the level.

**Comment by alakoring** [comment]
Since I fixed my code so it no longer logs errors, I have not seen this. But I was not implementing a logger.
I was using two loggers:
```	[DDLog addLogger: [DDASLLogger sharedInstance]];
	[DDLog addLogger: [DDTTYLogger sharedInstance]];```

**Comment by carllindberg** [comment]
The initial report did not post enough of the backtraces to see what the problem was.  It sure looked like one of those loggers was never completing, or another (unexpected) logger was not completing.  But without the backtrace it's impossible to tell.  It could get into Apple code which needs a separate resource which was already locked on another thread, for example.

**Comment by Elanz** [comment]
Yeah, The feed back here is correct for me.  Sync logging inside a log rolling process is a no-no.  I guess that doesn't count though as the same thing in the initial report.

Sorry for adding noise to the thread, carry on.

**Comment by wenssh** [comment]
We came up the same situation, how to fix this problem?

**Comment by ToroLiu** [comment]
I encounter this problem too. When I add multiple Loggers.

```
        // Apple System Logs
        DDLog.add(DDASLLogger.sharedInstance())
        // It's fine when add 2 logger.
        DDLog.add(fileLogger, with: .info)

        // This will lead to dead lock
        DDLog.add(CrashlyticsLogger.sharedInstance())

```

```
class CrashlyticsLogger : DDAbstractLogger
{
    static let sharedInstance = CrashlyticsLogger()
    
    override init() {
        super.init()
        
        // I have tried a custom formatter, it dead lock too.
        self.logFormatter = DDDispatchQueueLogFormatter()
    }
    
    override var loggerName: String! {
        return "CrashlyticsLogger"
    }

    override func log(message logMessage: DDLogMessage!) {
        guard let formatter = self.logFormatter else {
            NSLog("CrashlyticsLogger: No formatter")
            return
        }
        
        let msg = formatter.format(message: logMessage)!
        self.write(msg)
    }
    
    func write(_ string: String) {
        NSLog("1: \(string)")
        //CLSLogv("%@", getVaList([string]))
    }
}
```

     After calling `DDLogInfo`,  it dead lock when calling `DDLogError`.  If no calling `DDLogError`, nothing print out when executing `DDLog.add(CrashlyticsLogger.sharedInstance())`

**Comment by papley** [comment]
Still a big problem. I have 3 loggers and they end up deadlocking constantly. There are hints as to the problem here but this is far from a dead subject.

**Comment by ffried** [comment]
First of all, please make sure you don't call `self.logFormatter` inside `logMessage:` of a `DDAbstractLogger` subclass. This **will** lead to a deadlock. This is most certainly the issue in the code in @ToroLiu's comment (`guard let formatter = self.logFormatter ...`). In ObjC please access the instance variable directly (`self->_logFormatter`). In Swift, either get it via `class_getInstanceVariable` (as described [here](https://github.com/CocoaLumberjack/CocoaLumberjack/issues/643#issuecomment-156570032)) or use your own instance property (as described [here](https://github.com/CocoaLumberjack/CocoaLumberjack/issues/643#issuecomment-156981477)).

If that's not the case, can those of you who still experience this issue please post some updated stack traces? This would help to determine if the underlying issue is the same as in #910. And maybe even help to track it down.

**Comment by hhanesand** [comment]
Closing, since OP did not include a full backtrace. We're unable to do anything more at this point. Pinging @alakoring, @papley, and @wenssh. If you can reproduce this issue on master as of today, we'd love if you opened a new issue with a full backtrace (`bt all` in debugger) so we can investigate this issue further.
